apps/server/src/modules/oauth/service/oauth.service.ts

```diff
@@ -0,0 +1,159 @@
+		}
+		throw new OAuthSSOError('Authorization Query Object has no authorization code or error', errorCode);
+	}
+
+	async requestToken(code: string, system: System): Promise<OauthTokenResponse> {
+		const decryptedClientSecret: string = this.oAuthEncryptionService.decrypt(system.oauthConfig.clientSecret);
+		const tokenRequestPayload: TokenRequestPayload = TokenRequestPayloadMapper.mapToResponse(
+			system,
+			decryptedClientSecret,
+			code
+		);
+		const responseTokenObservable = this.httpService.post<OauthTokenResponse>(
+			tokenRequestPayload.tokenEndpoint,
+			QueryString.stringify(tokenRequestPayload.tokenRequestParams),
+			{
+				method: 'POST',
+				headers: {
+					'Content-Type': 'application/x-www-form-urlencoded',
+				},
+			}
+		);
+		const responseToken = await lastValueFrom(responseTokenObservable);
+		return responseToken.data;
+	}
+
+	async _getPublicKey(system: System): Promise<string> {
+		const client: JwksRsa.JwksClient = JwksRsa({
+			cache: true,
+			jwksUri: system.oauthConfig.jwksEndpoint,
+		});
+		const key: JwksRsa.SigningKey = await client.getSigningKey();
+		return key.getPublicKey();
+	}
+
+	async validateToken(idToken: string, system: System): Promise<IJWT> {
+		const publicKey = await this._getPublicKey(system);
+		const verifiedJWT = jwt.verify(idToken, publicKey, {
+			algorithms: ['RS256'],
+			issuer: system.oauthConfig.issuer,
+			audience: system.oauthConfig.clientId,
+		});
+		if (typeof verifiedJWT === 'string' || verifiedJWT instanceof String)
+			throw new OAuthSSOError('Failed to validate idToken', 'sso_token_verfication_error');
+		return verifiedJWT as IJWT;
+	}
+
+	async findUser(decodedJwt: IJWT, systemId: string): Promise<User> {
+		let user: User;
+		const system: System = await this.systemRepo.findById(systemId);
+		// iserv strategy
+		if (system.oauthConfig.provider === 'iserv') {
+			const uuid = this.iservOauthService.extractUUID(decodedJwt);
+			user = await this.iservOauthService.findUserById(uuid, systemId);
+			return user;
+		}
+		// TODO in general
+		// not working
+		user = await this.userRepo.findById(decodedJwt.sub);
+		return user;
+	}
+
+	async getJWTForUser(user: User): Promise<string> {
+		const jwtResponse: string = await this.jwtService.generateJwt(user.id);
+		return jwtResponse;
+	}
+
+	async processOauth(query: AuthorizationParams, systemId: string): Promise<OAuthResponse> {
+		const authCode = this.checkAuthorizationCode(query);
+		const system = await this.systemRepo.findById(systemId);
+		const queryToken = await this.requestToken(authCode, system);
+		const decodedToken = await this.validateToken(queryToken.id_token, system);
+		const user = await this.findUser(decodedToken, systemId);
+		const jwtResponse = await this.getJWTForUser(user);
+		const response: OAuthResponse = new OAuthResponse();
+		response.idToken = queryToken.id_token;
+		response.logoutEndpoint = system.oauthConfig.logoutEndpoint;
+		response.provider = system.oauthConfig.provider;
```

Comment:
L120-L123 could be a private method instead of an mapper

Reply:
done